### PR TITLE
Optimize synthetic deleted generated when replicating from storage

### PR DESCRIPTION
**Pull request by mlaflamm**
Do not generate child record delete when parent is deleted in the same transaction.

### CODE REVIEW

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala [hunk 1]
```diff
@@ -393,14 +392,21 @@ class MysqlStorage(config: MysqlStorageConfiguration, garbageCollection: Boolean
      */
     def applyTo(transaction: Transaction) {
       import com.wajam.mry.execution.Implicits._
+      import java.util
 
       // Set or delete each record
       val storage = transaction.from("mysql")
+      val deleted: util.NavigableSet[CompositeKey[(String, String)]] = new util.TreeSet()
```
Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala]: Why java TreeSet?

Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala]: http://docs.oracle.com/javase/7/docs/api/java/util/TreeSet.html#subSet(E, boolean, E, boolean)

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlTransaction.scala [hunk 2]
```diff
@@ -952,6 +952,25 @@ class MutationRecord {
   }
 }
 
+case class CompositeKey[T](keys: T*)(implicit ord: math.Ordering[T]) extends Comparable[CompositeKey[T]] {
```
Comment (coderunner) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlTransaction.scala:955]: Should we put this in another file? MysqlTransaction is already about 1k lines.

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala (hunk 1 above)
Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlStorage.scala]: Using a Scala sorted set

``` scala
deleted.dropWhile(_ < deleteKey.head).takeWhile(_ <= deleteKey.parent)
```

**Comment by mlaflamm** [comment]
I found a flaw in the solution when reflecting more about the usage of java TreeSet as raised by @alex-wajam. I will come back soon with something that actually works!

**Comment by alexbergeron** [comment]
Always my pleasure to make sure Java collections are used less and less

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlTransaction.scala (hunk 2 above)
Comment (mlaflamm) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlTransaction.scala:955]: I think we should move much more than this class if we want to go that way. For consistency sake, either we keep it in this file or also move AccessKey, AccessPath, Record, MutationRecord, RecordIterator, VersionRecord, etc...

**Comment by coderunner** [comment]
+1

**Comment by alexbergeron** [comment]
Makes a lot more sense. 

+1

File: mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlTransaction.scala (hunk 2 above)
Comment (alexbergeron) [inline mry-core/src/main/scala/com/wajam/mry/storage/mysql/MysqlTransaction.scala:955]: +1

**Pull request merged**